editoast/src/views/rolling_stocks.rs

```diff
@@ -167,10 +178,22 @@ async fn delete(db_pool: Data<DbPool>, path: Path<i64>) -> Result<HttpResponse>
     assert_rolling_stock_unlocked(
         retrieve_existing_rolling_stock(&db_pool, rolling_stock_id).await?,
     )?;
-    if !RollingStockModel::delete(db_pool, rolling_stock_id).await? {
-        return Err(RollingStockError::NotFound { rolling_stock_id }.into());
+    match get_rolling_stock_usage(db_pool.clone(), rolling_stock_id).await {
+        Ok(trains) => {
+            if trains.is_empty() {
+                return match RollingStockModel::delete(db_pool.clone(), rolling_stock_id).await {
+                    Ok(_) => Ok(HttpResponse::NoContent().finish()),
+                    Err(err) => Err(err),
+                };
+            }
+            Err(RollingStockError::RollingStockIsUsed {
+                rolling_stock_id,
+                usage: trains,
+            }
+            .into())
+        }
+        Err(err) => Err(err),
     }
```

Comment:
Why not:
```suggestion
    let trains = get_rolling_stock_usage(db_pool.clone(), rolling_stock_id).await?;
    if trains.is_empty() {
        return match RollingStockModel::delete(db_pool.clone(), rolling_stock_id).await {
            Ok(_) => Ok(HttpResponse::NoContent().finish()),
            Err(err) => Err(err),
        };
    }
    Err(RollingStockError::RollingStockIsUsed {
        rolling_stock_id,
        usage: trains,
    }
    .into())
```
This would avoid the extra nesting

Also, you removed the ` Err(RollingStockError::NotFound { rolling_stock_id }.into());`, is it voluntary ?

Reply:
```
    if !RollingStockModel::delete(db_pool, rolling_stock_id).await? {
        return Err(RollingStockError::NotFound { rolling_stock_id }.into());
```
That was removed because the check is now done by the `get_rolling_stock_usage`

Reply:
good stuff, i took that change :)